### PR TITLE
Fix #1542: Use file streaming instead of reading to memory

### LINKED ISSUE #1542: Use file streaming instead of reading to memory
Currently, the API reads file into memory on upload _and_ `/assets/` download, which makes it impossible to upload files that are bigger than ~100MB* and makes opening existing thumbnails slow.

We should use Streams to read / write / serve files, instead of reading them into memory whenever we can.

https://flysystem.thephpleague.com/docs/usage/filesystem-api/
https://odan.github.io/2018/03/09/emitting-chunked-stream-responses-with-slim.html

\* Depending on your server's memory and max php limits

### PR DESCRIPTION
**Pull request by alecreed**
Fixes #1542. Adds an optional boolean parameter $asStream (default false) to several methods on AssetService and makes use of that parameter in the Assets route. I verified locally with memory_get_peak_usage while serving the original of a large image that the file was no longer being loaded into memory in its entirety.

(This is my first pull request ever.)

### CODE REVIEW

**Comment by alecreed** [comment]
My mistake, #1542 also complains of files being fully loaded into memory on upload as well. Second commit also fixes that part. (Should also fix #1729 as well, since I was editing that method heavily and the relevant code stood out).

Basically, Filesystem\Files was making an on-disk copy of the temporary file created by php's built-in upload handling, then (for images at least) loading that copy into memory in order to call getimagesizefromstring on the actual bytes to extract metadata. I changed it to use the existing temporary file and to call getimagesize instead, which I believe only reads a minimal header from the file. Memory usage now looks stable regardless of the size of the uploaded file.

I only tested using the app and running on my local filesystem (not s3), so it's possible there are corner cases I'm not aware of.

**Comment by everyx** [comment]
@alecreed If I'm free next week, I'll test s3 compatibility :)

**Comment by rijkvanzanten** [comment]
This is super awesome. Thanks @alecreed! 

Lemme know what you find @everyx! I'd like to make sure we don't run into any unforeseen side effects when merging this 🙂

File: src/core/Directus/Filesystem/Files.php [hunk 1]
```diff
@@ -406,13 +386,19 @@ public function saveEmbedData(array $fileInfo)
...
      * @param string $path - file path
      * @param bool $outside - if the $path is outside of the adapter root path.
+     * @param string|null $localCopy - path to local copy of the file, if available
      *
      * @throws \RuntimeException
      *
      * @return array file information
      */
-    public function getFileInfo($path, $outside = false)
+    public function getFileInfo($path, $outside = false, $localCopy = null)
```
Comment (WoLfulus) [inline src/core/Directus/Filesystem/Files.php]: shouldn't this become a reference? &$localCopy since we check for $tmp on the `saveData` function?

Also, shouldn't we check for $tmp with `is_null` instead of isset?

Comment (alecreed) [reply src/core/Directus/Filesystem/Files.php]: On the reference, I don't think so. It should always be a string path to a file (if it's not null), and it isn't modified inside getFileInfo or getFileInfoFromLocalCopy.

Do you mean `!is_null`? We only want to run the `ffprobe` if `$tmp` is a string path to a file (I think). I added that early on because previously it was possible for `$tmp` [to be undefined](https://github.com/directus/api/blob/5a21686fdb7b456ec8c8999ac8b3ea372c6a55e3/src/core/Directus/Filesystem/Files.php#L321-L326) at that point in the function. The fact it never seems to have caused runtime errors makes me suspect that `saveData` is currently only being used for file uploads (i.e. with an `UploadedFile` for the `$fileData` argument) even though it looks like it was written with an eye toward supporting a second option where `$fileData` would be a string of some sort (see some of the comments referencing base64 data, which I didn't change).

I'm new to the codebase, though, so I'm not sure if that's accurate and I wanted to err on the side of making fewer changes if I could.

**Comment by everyx** [comment]
> This is super awesome. Thanks @alecreed!
> 
> Lemme know what you find @everyx! I'd like to make sure we don't run into any unforeseen side effects when merging this 🙂

Sure, so busy now, I'll test this weekend. :)

**Comment by everyx** [comment]
@rijkvanzanten Just tested, it works fine with s3. :)

**Comment by everyx** [comment]
@rijkvanzanten Can we merge this after @alecreed fix the conflicting?

**Comment by rijkvanzanten** [comment]
I think so!

**Comment by alecreed** [comment]
First conflict I've ever resolved. I think I did it right. It wasn't anything material to this change so much as an alternate way to fix a different issue. Should I resolve the @WoLfulus conversation or is that rude?

**Comment by rijkvanzanten** [comment]
That's OK! As long as you don't resolve any open requests or comments without.. resolving them (either in code or in text) it isn't rude 🙂

File: src/core/Directus/Filesystem/Files.php (hunk 1 above)
Comment (alecreed) [reply src/core/Directus/Filesystem/Files.php]: I think we're good on this. Feel free to let me know of any other concerns.

**Comment by rijkvanzanten** [comment]
I know it took a minute for me to review it, but @alecreed this is phenomenal! Thanks a lot ❤️
